### ISSUE TITLE
AI-generated audit: potential undefined behavior and data races in Metal kernels

### ISSUE THREAD

**Issue by norpadon** [trymirai/uzu#523]
> **AI-generated report:** This issue was drafted by an AI assistant from a static audit of the Metal kernels. Please treat it as a review aid, not as a definitive diagnosis. Each item below should be validated with targeted tests or a small repro.

## Summary

I found several places where Metal kernels appear to rely on contracts that are not fully enforced by the Rust launch/configuration code. The main classes are:

- Out-of-bounds threadgroup or device memory access.
- Reads from uninitialized scratch/partial buffers.
- Misaligned vector pointer loads after casting scalar pointers to `vec<T, 4>*`.
- Non-atomic concurrent writes to the same device memory slot.
- Missing shape/configuration checks before division or indexing.

This report intentionally explains the Metal/C++ memory-safety concerns in plain terms. A few references that are useful background:

- Apple documents `MTLPointerType.alignment` as the required byte alignment for pointer element data: https://developer.apple.com/documentation/metal/mtlpointertype/alignment
- C++ object types have alignment requirements; creating/accessing an object at storage that does not meet its alignment requirement is undefined behavior: https://eel.is/c++draft/basic.align
- `reinterpret_cast` between object pointer types does not remove alignment/type-access constraints: https://eel.is/c++draft/expr.reinterpret.cast and https://eel.is/c++draft/basic.lval
- C++ data races are conflicting non-atomic concurrent actions without a happens-before relation, and any such race is undefined behavior: https://eel.is/c++draft/intro.races

## Navigation

- [High-confidence findings](#high-confidence-findings)
  - [1. MoE counts scratch buffer OOB](#1-moe-counts-scratch-buffer-oob)
  - [2. MoE partials layout mismatch](#2-moe-partials-layout-mismatch)
  - [3. MoE router `x_cache` overflow](#3-moe-router-x_cache-overflow)
  - [4. MoE tile scan under-covers experts](#4-moe-tile-scan-under-covers-experts)
  - [5. MoE 1D gather duplicate writes](#5-moe-1d-gather-duplicate-writes)
  - [6. Quantized GEMM B-loader tail masking](#6-quantized-gemm-b-loader-tail-masking)
  - [7. Misaligned vector loads](#7-misaligned-vector-loads)
  - [8. Hadamard tail OOB](#8-hadamard-tail-oob)
  - [9. Sampling can read `logits[UINT_MAX]`](#9-sampling-can-read-logitsuint_max)
- [Medium-confidence / contract-leak findings](#medium-confidence--contract-leak-findings)
  - [10. Repetition penalty duplicate-token race](#10-repetition-penalty-duplicate-token-race)
  - [11. Attention `QkUnpack` missing shape contract](#11-attention-qkunpack-missing-shape-contract)
  - [12. U4 quantized embedding odd dimension](#12-u4-quantized-embedding-odd-dimension)
- [Suggested validation tests](#suggested-validation-tests)

## Finding summary

| Severity | Area | Issue |
|---|---|---|
| High | MoE | Counts scratch buffer OOB for `num_routed_experts > 128` |
| High | MoE | Per-block partials are read as if initialized, but only global counts are written |
| High | MoE | Router `x_cache` overflows for `d_model > 4096` |
| High | MoE | Tile scan launches 256 threads but supports up to 512 experts |
| High | MoE | 1D gather ignores `tgid_x`, causing duplicate writes |
| High | GEMM | Quantized B tail masking appears to use swapped dimensions |
| High | GEMV/audio/SSD | Several vectorized loads can be misaligned |
| High | Hadamard | Tail lanes read/write past `hidden_dim` |
| High | Sampling | Empty candidate set can load `logits[UINT_MAX]` |
| Medium | Sampling | Repetition penalty can race on duplicate token ids |
| Medium | Attention | `QkUnpack` assumes valid `num_groups`/GQA divisibility |
| Medium | Embedding | U4 embedding odd `model_dim` reads past packed row |

### 1. MoE counts scratch buffer OOB

`counts_shared` is declared with 128 entries:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/counts_offsets_fused.metal#L54

But the tiled expert loop writes `counts_shared[e0 + e]`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/counts_offsets_fused.metal#L100

The host-side constructor allows up to 512 routed experts:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/encodable_block/mlp/moe/mod.rs#L97

So for `num_routed_experts > 128`, the kernel can index past the 128-entry threadgroup array. Threadgroup memory is local to the threadgroup, but it is still real storage with fixed bounds; indexing past the declared array can corrupt neighboring threadgroup variables or produce nonsensical scan results.

Possible fixes:

- Size `counts_shared` for the maximum supported expert count.
- Scan counts in 128-expert chunks without indexing the scratch buffer by global expert id.
- Reduce the host contract to `num_routed_experts <= 128`.

### 2. MoE partials layout mismatch

The counts kernel writes one count per expert:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/counts_offsets_fused.metal#L101

But `MoeBlockBasesFromPartials` reads `partials` as if it has a per-block layout:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/scatter_buckets.metal#L151

The Rust caller allocates `partials` for `num_blocks * num_tiles * 512` entries:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/encodable_block/mlp/moe/mod.rs#L245

However, the current writer only initializes the first tile/global-count region, not every block region. When `num_blocks > 1`, later block entries are read before being written.

Possible fixes:

- Make `MoeCountsOffsetsFused` write true per-block partials.
- Or change block-base computation to consume the actual global-count layout.
- Add tests with `batch_dim > 256`, since that is where `num_blocks > 1`.

### 3. MoE router `x_cache` overflow

The router declares a fixed 1024-entry `float4` cache:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/router_topk.metal#L26

It computes `vecs = d_model / 4` and writes:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/router_topk.metal#L42
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/router_topk.metal#L46

Any `d_model > 4096` means `vecs > 1024`, so the cache overflows. The constructor checks only that `model_dim % 4 == 0`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/encodable_block/mlp/moe/mod.rs#L91

Possible fixes:

- Enforce `model_dim <= 4096` for this kernel.
- Stream over `d_model` rather than caching the whole row.
- Cache fixed-size chunks.

### 4. MoE tile scan under-covers experts

`MoeTileScan` comments that `scratch[1024]` supports up to `E <= 1023`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/tiles_map.metal#L28

But the kernel launches only 256 threads:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/tiles_map.metal#L29

The load and scan only happen for `tid < e`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/tiles_map.metal#L31

For `e > 256`, entries above 255 are never loaded or scanned. Later code reads `scratch[e - 1]` and `tile_row_offsets[gid]`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/tiles_map.metal#L62
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/tiles_map.metal#L86

Possible fixes:

- Launch 1024 threads for this scan.
- Or replace it with a multi-pass scan.
- Or lower the host maximum to `num_routed_experts <= 256` for this path.

### 5. MoE 1D gather duplicate writes

`MoeGatherXPerm1D` dispatches multiple threadgroups:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/gather.metal#L81

But the work loop starts at `row = lid`, ignoring `tgid_x`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/gather.metal#L91

That means every dispatched threadgroup copies the same rows and writes the same `x_perm` addresses:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/gather.metal#L103
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/moe/gather.metal#L109

Even if the value is intended to be identical, non-atomic concurrent writes are still a race pattern and are not a safe contract to rely on.

Possible fix:

```metal
for (uint row = tgid_x * THREADS_PER_THREADGROUP + lid;
     row < total_rows;
     row += THREADS_PER_THREADGROUP * grid_threadgroup_count) {
  ...
}
```

Or dispatch only one threadgroup if this kernel is intended to be single-group.

### 6. Quantized GEMM B-loader tail masking

The quantized B loaders check `tile_row_index` against `src_tile_dim.x` when `REDUCTION_DIMENSION == 1`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/matmul/gemm/common/quant_scale_bias.h#L96
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/matmul/gemm/common/quant_scale_zero_point.h#L133

But callers pass B tile dimensions as `(K, N)` for transposed B tiles:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/matmul/gemm/common/simdgroup_mma_core.h#L111
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/matmul/gemm/common/mxu_mma_core.h#L112

That appears to check the row index against the K dimension instead of the N/tile-row dimension. In tail-N tiles, rows beyond valid `N` may still load `scales`, `biases`, `zero_points`, and packed weights.

Possible fixes:

- Audit and document the `short2` convention for B-loader `load_safe`.
- For transposed quantized B, validate `tile_row_index` against the remaining N rows and pack/column indices against the remaining K columns.
- Add tests for quantized GEMM with non-multiple `N` and non-multiple `K`.

### 7. Misaligned vector loads

Several kernels cast scalar pointers to vector pointers and dereference them. This is only safe if the address is aligned for the vector type. A `reinterpret_cast` does not make an address aligned; it only changes the static pointer type.

Concrete cases:

Full-precision GEMV:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/matmul/gemv/common/full_precision_b_source.h#L33
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/matmul/gemv/common/full_precision_b_source.h#L37

The selector can still choose this path with `input_aligned == false`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/matmul/gemv/kernel.rs#L86
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/matmul/gemv/kernel.rs#L92

Audio grouped conv:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/audio/causal_conv1d_grouped.metal#L84
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/audio/causal_conv1d_grouped_residual.metal#L92

Here, `x_t = t + k * dilation - pad`, so even if `t` is tile-aligned, dilation/padding can make the actual source address non-4-aligned.

SSD sequential prefill:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/ssm/ssd_prefill.metal#L251
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/ssm/ssd_prefill.metal#L252
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/ssm/ssd_prefill.metal#L253

Later rows can be misaligned if `state_dim` or the relevant strides are not multiples of 4.

Possible fixes:

- Only use vector loads when the scalar element index is a multiple of the vector width and the base pointer alignment is guaranteed.
- Otherwise use scalar loads.
- Encode the alignment contract in Rust-side shape validation when vectorization is required.

### 8. Hadamard tail OOB

`HadamardTransform` dispatches `hidden_dim.div_ceil(METAL_SIMD_SIZE)`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/hadamard_transform/hadamard_transform.metal#L18

But it unconditionally reads and writes `data[element_index]` and `factors[factor_index]`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/hadamard_transform/hadamard_transform.metal#L26
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/hadamard_transform/hadamard_transform.metal#L30

If `hidden_dim` is not a multiple of 32, tail lanes in the final block access past the row.

Possible fixes:

- Enforce `hidden_dim % 32 == 0` wherever Hadamard factors are accepted.
- Or guard tail lanes in the kernel.

### 9. Sampling can read `logits[UINT_MAX]`

`UnifiedSampling` initializes the reduced candidate to `Logit::LOWEST`, whose index is `UINT_MAX`, then later loads:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/sampling/unified_sampling.metal#L34
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/sampling/unified_sampling.metal#L164

If the candidate set is empty, the identity value can survive the reduction. Examples include:

- `vocab_size == 0`
- `top_k == 0`
- all tokens masked out by `bitmask`

Possible fixes:

- Validate sampling parameters before launch.
- Reject or define behavior for an empty candidate set.
- Add a kernel-side guard before loading the candidate index.

### 10. Repetition penalty duplicate-token race

`RepetitionPenalty` writes the current token:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/sampling/repetition_penalty.metal#L20

And also writes ring tokens:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/sampling/repetition_penalty.metal#L31
https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/sampling/repetition_penalty.metal#L36

If the current token appears in the ring, or if the ring contains duplicates, multiple threads can write the same `logits_copy[token_id]`. The written value may be identical, but conflicting non-atomic writes are still a race pattern.

Possible fixes:

- Deduplicate token ids before applying the penalty.
- Serialize this small operation.
- Assign a unique owner per token id.

### 11. Attention `QkUnpack` missing shape contract

`QkUnpack` divides by `num_groups`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/attention/qk_unpack.metal#L19

It also assumes `num_heads % num_groups == 0`; otherwise `group_index` can reach `num_groups` and write past `unpacked_keys`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/attention/qk_unpack.metal#L34

The CPU implementation appears to check this contract, but the Metal path should enforce it before launch too.

Possible fix:

- Add constructor/config validation for `num_groups != 0` and `num_heads.is_multiple_of(num_groups)`.

### 12. U4 quantized embedding odd dimension

For U4, `weights_stride` is `model_dim / 2`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/embedding/quant_embedding.metal#L45

The last odd dimension reads `dim_idx / 2`, which equals `model_dim / 2` for odd `model_dim`:

https://github.com/trymirai/uzu/blob/main/crates/backend-uzu/src/backends/metal/kernel/embedding/quant_embedding.metal#L50

That is one byte past a floor-packed row.

Possible fixes:

- Require even `model_dim` for U4 embeddings.
- Or use ceil packing consistently in validation and kernel indexing.

## Suggested validation tests

- MoE:
  - `num_routed_experts = 129, 257, 512`
  - `batch_dim > 256`
  - `model_dim = 4100`
- GEMM:
  - Quantized B with tail `N`
  - Quantized B with tail `K`
  - Tail `N` and tail `K` together
- GEMV:
  - Full-precision `K = 129` or another non-4-aligned value that still selects GEMV
- Audio grouped conv:
  - A dilation/padding combination where the fast path uses `x_t % 4 != 0`
- Hadamard:
  - `hidden_dim = 33`
- Sampling:
  - `top_k = 0`
  - empty bitmask
  - `vocab_size = 0`, if constructible
- Attention:
  - `num_groups = 0`
  - `num_heads` not divisible by `num_groups`
- U4 embedding:
  - odd `model_dim`

**Comment by uuuvn** [comment]
Every single one I can quickly judge is utter slop, @eugenebokhan can you quickly skip the gemm ones in case they are real? Probably aren't but worth a quick glance
